File: 19-RemoveNthNodeFromEndOfList/19-RemoveNthNodeFromEndOfList.py
# Last updated: 6/11/2026, 11:38:34 AM
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if(head is None):
            return
        # curr = head
        # length = 0
        
        # while(curr):
        #     curr = curr.next
        #     length += 1

        # curr = head
        # cn = 0
        # gn = length - n + 1
        # if(length == n):
        #     return head.next
        # while(curr):
        #     cn += 1
        #     if(cn==gn-1):
        #         curr.next = curr.next.next
        #         return head
        #     curr = curr.next
        dummy = ListNode(0,head)
        first = dummy
        second = head
        
        while(n>0):
            second = second.next
            n -= 1

        while(second):
            first = first.next
            second = second.next

        first.next = first.next.next
        return dummy.next


        



        
            

File: 19-RemoveNthNodeFromEndOfList/19-RemoveNthNodeFromEndOfList.cpp
// Last updated: 6/11/2026, 11:38:28 AM
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if(head == nullptr || head->next == nullptr) return nullptr;
        ListNode* cur = head;
        ListNode* cur2 = head;

        int ct = 0;

        while(ct<n){
            cur2 = cur2->next;
            ct++;
        }

        while(cur2 && cur2->next){
            cur = cur->next;
            cur2 = cur2->next;
        }

        if(cur2 == nullptr){
            return head->next;
        }

        cur->next = cur->next->next;
        return head;

    }
};